aya/src/maps/ringbuf.rs

```diff
@@ -0,0 +1,338 @@
+//! programs to userspace.
+//!
+//! [ringbuf]: https://www.kernel.org/doc/html/latest/bpf/ringbuf.html
+
+use std::{
+    io,
+    ops::Deref,
+    os::unix::prelude::AsRawFd,
+    ptr,
+    sync::atomic::{fence, AtomicU32, AtomicUsize, Ordering},
+};
+
+use libc::{munmap, sysconf, MAP_FAILED, MAP_SHARED, PROT_READ, PROT_WRITE, _SC_PAGESIZE};
+
+use crate::{
+    generated::{BPF_RINGBUF_BUSY_BIT, BPF_RINGBUF_DISCARD_BIT, BPF_RINGBUF_HDR_SZ},
+    maps::{MapData, MapError},
+    sys::mmap,
+};
+
+/// A map that can be used to receive events from eBPF programs.
+///
+/// This is similar to [`PerfEventArray`], but different in a few ways:
+/// * It's shared across all CPUs, which allows a strong ordering between events. It also makes the
+///   buffer creation easier.
+/// * Data notifications are delivered for every event instead of being sampled for every N event;
+///   the eBPF program can also control notification delivery if sampling is desired for performance reasons.
+/// * On the eBPF side, it supports the reverse-commit pattern where the event can be directly
+///   written into the ring without copying from a temporary location.
+/// * Dropped sample notifications goes to the eBPF program as the return value of `reserve`/`output`,
+///   and not the userspace reader. This might require extra code to handle, but allows for more
+///   flexible schemes to handle dropped samples.
+///
+/// To receive events you need to:
+/// * call [`RingBuf::try_from`]
+/// * poll the returned [`RingBuf`] to be notified when events are inserted in the buffer
+/// * call [`RingBuf::next`] to read the events
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 5.8.
+///
+/// # Examples
+///
+/// The following example shows how to read samples as well as using an async runtime
+/// to wait for samples to be ready:
+///
+/// ```no_run
+/// # use aya::maps::{Map, MapData, RingBuf};
+/// # use std::ops::DerefMut;
+/// # #[derive(thiserror::Error, Debug)]
+/// # enum Error {
+/// #    #[error(transparent)]
+/// #    IO(#[from] std::io::Error),
+/// #    #[error(transparent)]
+/// #    Map(#[from] aya::maps::MapError),
+/// #    #[error(transparent)]
+/// #    Bpf(#[from] aya::BpfError),
+/// # }
+/// # struct Poll<T: AsRef<MapData>>(RingBuf<T>);
+/// # impl<T: AsRef<MapData>> Poll<T> {
+/// #    fn new(inner: RingBuf<T>) -> Self { Self (inner) }
+/// #    async fn readable(&mut self) {}
+/// #    fn get_inner_mut(&mut self) -> &mut RingBuf<T> { &mut self.0 }
+/// # }
+/// # async {
+/// # let mut bpf = aya::Bpf::load(&[])?;
+/// use std::convert::{TryFrom, TryInto};
+///
+/// let mut ring = RingBuf::try_from(bpf.map_mut("EVENTS").unwrap())?;
+///
+/// // Poll would be a helper that takes an `AsRawFd` implementation and provides functionality
+/// // to wait for the file descriptor to be readable.
+/// let mut poll = Poll::new(ring);
```

Comment:
I don't think it's a good example of how using ring buffer with Tokio should look like. IMO we shouldn't require users to write or provide any wrapper structures just to consume events. Also, it's quite inconsistent with the API we have for perf buffers with `AsyncPerfEventArray`.

I was playing with that branch recently and I think that it would be nice to have a dedicated "async" struct, similar to `AsyncPerfEventArray` . My idea is to have some method which consumes a bytes buffer and fills it with data from ring buffer, without returning the `RingBufItem` - so the user would just get the data and the original ring buffer item would be dropped, therefore immediately updating producer/consumer positions.  I imagine that it could look somehow like:

```rust
use std::os::{fd::AsRawFd, unix::prelude::RawFd};

#[cfg(feature = "async_std")]
use async_io::Async;
use bytes::BytesMut;
use thiserror::Error;
#[cfg(feature = "async_tokio")]
use tokio::io::unix::AsyncFd;

use crate::maps::{ringbuf::RingBuf, MapData, MapError};

/// Async ring buffer error.
#[derive(Error, Debug)]
pub enum AsyncRingBufError {
    /// A map error.
    #[error("map error: {0}")]
    MapError(MapError),

    /// An IO error.
    #[error(transparent)]
    IOError(#[from] std::io::Error),
}

#[cfg(any(feature = "async_std", feature = "async_tokio"))]
pub struct AsyncRingBuf<T: AsRef<MapData>> {
    ring_buf: RingBuf<T>,

    #[cfg(feature = "async_std")]
    async_fd: Async<RawFd>,

    #[cfg(feature = "async_tokio")]
    async_fd: AsyncFd<RawFd>,
}

#[cfg(any(feature = "async_std", feature = "async_tokio"))]
impl<T: AsRef<MapData>> AsyncRingBuf<T> {
    pub(crate) fn new(map: T) -> Result<AsyncRingBuf<T>, MapError> {
        let ring_buf = RingBuf::new(map)?;
        #[cfg(any(feature = "async_std", feature = "async_tokio"))]
        let fd = ring_buf.as_raw_fd();
        Ok(AsyncRingBuf {
            ring_buf: ring_buf,
            #[cfg(feature = "async_std")]
            async_fd: Async::new(fd).unwrap(),
            #[cfg(feature = "async_tokio")]
            async_fd: AsyncFd::new(fd).unwrap(),
        })
    }
}

#[cfg(feature = "async_tokio")]
impl<T: AsRef<MapData>> AsyncRingBuf<T> {
    /// Try to take a new entry from the ringbuf.
    pub async fn read_item(&mut self, buffer: &mut BytesMut) -> Result<usize, AsyncRingBufError> {
        loop {
            let mut guard = self.async_fd.readable_mut().await?;
            match self.ring_buf.next().as_deref() {
                Some(item) => {
                    buffer.extend_from_slice(item);
                    return Ok(item.len());
                }
                None => {
                    guard.clear_ready();
                    continue;
                }
            }
        }
    }
}
```

And the userspace code consuming it could look like:

```rust
let mut ring = AsyncRingBuf::try_from(bpf.map_mut("EVENTS").unwrap()).unwrap();
let mut buf = BytesMut::with_capacity(1024);

loop {
    let _ = ring.read_item(&mut buf).await.unwrap();

    // Here I'm just using u64 for simplicity, but we could handle any type.
    let val = u64::from_ne_bytes(buf[..mem::size_of::<u64>()].try_into().unwrap());

    info!("{}", val);

    buf.clear();
}
```

(I didn't include `read_item` implementation for async-std in this example, but we should have it too)

/cc @alessandrod

Reply:
OK, actually my example is bad, because it always copies the item. I will try to come up with a better one which doesn't drop the item.

Reply:
Just curious: would "leave out the async integration" mean that reading the ring buffer from within a dedicated tokio task (thread) wouldn't be supported?

Reply:
No, it should work, just that putting that in a wrapper function happens to be tripping compile error for no reason. I do it like this:

```rust
    let ring_processor = async move {
        // Throttle polling to get batched and more CPU-efficient event processing
        let interval = Duration::from_micros(200);
        let mut next_poll = Instant::now() + interval;
        loop {
            let time_before_poll = Instant::now().saturating_duration_since(next_poll);
            if !time_before_poll.is_zero() {
                sleep(time_before_poll).await;
                next_poll = next_poll + interval;
            } else {
                next_poll = Instant::now() + interval;
            }
            let mut guard = ring.readable_mut().await?;
            while let Some(e) = guard.get_inner_mut().next() {
                // do something with e
            };
            guard.clear_ready();
        }
        Ok::<(), anyhow::Error>(())
    };

```

Reply:
Ok, thanks. I'll leave those details to you guys. Just wanted to chime in and say that being able to process a ring buffer's events in a dedicated/separate tokio task is a necessity, at least for my use case :)